### PR TITLE
Rdrp 858 mapper config validation

### PR DESCRIPTION
**Pull request by roodm27**
## Pull Request submission
Added function which checks if any mappers are empty or missing, checks the files are .csv and the survey year matches the mapper file names. 

 
*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/utils/config.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 from src.utils.defence import type_defence, validate_file_extension
 from src.utils.local_file_mods import safeload_yaml
-from src.utils.path_helpers import update_config_with_paths
+from src.utils.path_helpers import update_config_with_paths, validate_mapping_filenames
```
Comment (AnneONS) [inline src/utils/config.py:7]: this can go

File: src/pipeline.py [hunk 2]
```diff
@@ -67,7 +67,8 @@ def run_pipeline(user_config_path, dev_config_path):
 
     MainLogger.info("Launching Pipeline .......................")
     logger.info("Collecting logging parameters ..........")
-
+    
+    config = validate_mapping_filenames(config) #
```
Comment (AnneONS) [inline src/pipeline.py]: this could be done in the path_helpers script

File: src/mapping/mapping_helpers.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 """Specific functions applied in mapping_main.py"""
 import pandas as pd
 import logging
+from src.utils.defence import validate_file_extension
```
Comment (AnneONS) [inline src/mapping/mapping_helpers.py]: we don't need this now

File: src/pipeline.py [hunk 4]
```diff
@@ -16,6 +16,7 @@
 from src.estimation.estimation_main import run_estimation
 from src.site_apportionment.site_apportionment_main import run_site_apportionment
 from src.outputs.outputs_main import run_outputs
+from src.utils.path_helpers import validate_mapping_filenames
```
Comment (AnneONS) [inline src/pipeline.py]: this can go

File: src/pipeline.py [hunk 5]
```diff
@@ -67,7 +67,8 @@ def run_pipeline(user_config_path, dev_config_path):
 
     MainLogger.info("Launching Pipeline .......................")
     logger.info("Collecting logging parameters ..........")
-
+    
+    #config = validate_mapping_filenames(config) #
```
Comment (AnneONS) [inline src/pipeline.py]: this can go

File: src/utils/path_helpers.py [hunk 6]
```diff
@@ -167,6 +170,46 @@ def create_exports_config(config: dict) -> dict:
...
+    """Checks that the mapping filenames are valid"""
+    bool_dict, msg = validate_mapping_filenames(config)
+
+    if all(bool_dict.values()):
+        PathHelpLogger.info("All mapping filenames are valid.")
+    else:
+        PathHelpLogger.error("There are errors with the mapping filenames.")
+        raise ValueError(msg)
+        
+    return config
```
Comment (AnneONS) [inline src/utils/path_helpers.py:216]: looks good!

File: src/utils/path_helpers.py [hunk 7]
```diff
@@ -186,4 +229,7 @@ def update_config_with_paths(config: dict, modules: list) -> dict:
     for module_name in modules:
         config[f"{module_name}_paths"] = create_module_config(config, module_name)
 
-    return config
+        ## add in function below
```
Comment (AnneONS) [inline src/utils/path_helpers.py]: maybe add a brief comment that says what the new function is doing?

File: tests/test_mapping/test_mapping_helpers.py [hunk 8]
```diff
@@ -250,3 +251,68 @@ def test_create_additional_ni_cols(self):
...
+    def test_validate_mapper_config_raises_file_incorrect(self):
+        config = {
+            'years' : {'survey_year': 2022,},
+            '2022_mappers': {'itl_mapper_path': "itl_2022.cs",
+            'ultfoc_mapper_path': "BERD_2022_ultfoc.csv",},
+            '2023_mappers': {'itl_mapper_path': "itl_2023.csv",
+            'ultfoc_mapper_path': None,}
+            } 
+
+        bool_dict = {}
```
Comment (AnneONS) [inline tests/test_mapping/test_mapping_helpers.py]: this isn't needed here, also below

File: tests/test_mapping/test_mapping_helpers.py [hunk 9]
```diff
@@ -250,3 +251,68 @@ def test_create_additional_ni_cols(self):
...
+            'years' : {'survey_year': 2022,},
+            '2022_mappers': {'itl_mapper_path': "itl_202.csv",
+            'ultfoc_mapper_path': "BERD_2022_ultfoc.csv",},
+            '2023_mappers': {'itl_mapper_path': "itl_2023.csv",
+            'ultfoc_mapper_path': None,}
+            } 
+
+        bool_dict = {}
+
+        validate_mapping_filenames(config)
```
Comment (AnneONS) [inline tests/test_mapping/test_mapping_helpers.py]: what does this function return?

you need to have 
expected_result, expected_msg = validate_mapping_filenames(config)

**Review by AnneONS — COMMENTED** [review]
Hi Matt, it's looking good! The tests need a bit more work though

**Review by zorge69 — APPROVED** [review]

**Pull request merged**